File: examples/tito_example.swift
// Together In Together Out - Swift example.
//
// Run: swift examples/tito_example.swift

/// Raised when an operation would violate the TITO invariants.
enum TitoError: Error, CustomStringConvertible {
    case emptyGroup(String)
    case duplicateMembers(String)
    case alreadyQueued(String)
    case membersQueued(String)
    case memberNotQueued(String)
    case notAMember(member: String, group: String)
    case groupNotQueued(String)

    var description: String {
        switch self {
        case .emptyGroup(let id):
            return "group \(id) must contain at least one member"
        case .duplicateMembers(let id):
            return "group \(id) contains duplicate members"
        case .alreadyQueued(let id):
            return "group \(id) is already in the queue"
        case .membersQueued(let members):
            return "members already in the queue: \(members)"
        case .memberNotQueued(let member):
            return "\(member) is not in the queue"
        case .notAMember(let member, let group):
            return "\(member) is not a member of group \(group)"
        case .groupNotQueued(let id):
            return "group \(id) is not in the queue"
        }
    }
}

/// A set of members admitted together. A reference type, so the group a caller
/// holds stays in step with the copy the queue tracks. It departs only once
/// every member has been marked ready.
final class Group: CustomStringConvertible {
    let groupID: String
    let members: [String]
    private var readyFlags: [String: Bool]
    private(set) var readyCount = 0

    init(_ groupID: String, _ members: [String]) throws {
        if members.isEmpty {
            throw TitoError.emptyGroup(groupID)
        }
        var flags: [String: Bool] = [:]
        for member in members {
            if flags[member] != nil {
                throw TitoError.duplicateMembers(groupID)
            }
            flags[member] = false
        }
        self.groupID = groupID
        self.members = members
        self.readyFlags = flags
    }

    /// Members already marked ready, in admission order.
    var readyMembers: [String] {
        return members.filter { readyFlags[$0] == true }
    }

    /// Members not yet marked ready, in admission order.
    var waitingMembers: [String] {
        return members.filter { readyFlags[$0] == false }
    }

    var isReady: Bool {
        return readyCount == members.count
    }

    /// Mark one member ready. Returns true once the whole group is ready.
    @discardableResult
    func markReady(_ member: String) throws -> Bool {
        guard let ready = readyFlags[member] else {
            throw TitoError.notAMember(member: member, group: groupID)
        }
        if !ready {
            readyFlags[member] = true
            readyCount += 1
        }
        return isReady
    }

    var count: Int {
        return members.count
    }

    func contains(_ member: String) -> Bool {
        return readyFlags[member] != nil
    }

    var description: String {
        return "\(groupID) [\(members.joined(separator: " "))]"
    }
}

/// A queue that admits and releases members group by group.
///
/// `strictOrder` true (the default policy) is head-of-line FIFO: only the
/// oldest waiting group may depart. `strictOrder` false lets any fully ready
/// group depart, still oldest first, so a waiting group does not block the
/// ready groups behind it.
final class TitoQueue {
    let strictOrder: Bool
    private var order: [Group] = []
    private var byID: [String: Group] = [:]
    private var memberIndex: [String: String] = [:]

    init(strictOrder: Bool = true) {
        self.strictOrder = strictOrder
    }

    /// Admit members as one group and return it.
    @discardableResult
    func admit(_ groupID: String, _ members: [String]) throws -> Group {
        let group = try Group(groupID, members)
        if byID[groupID] != nil {
            throw TitoError.alreadyQueued(groupID)
        }
        let queued = group.members.filter { memberIndex[$0] != nil }
        if !queued.isEmpty {
            throw TitoError.membersQueued(queued.joined(separator: " "))
        }
        order.append(group)
        byID[groupID] = group
        for member in group.members {
            memberIndex[member] = groupID
        }
        return group
    }

    /// Mark a queued member ready. Returns true once its group is ready.
    @discardableResult
    func markReady(_ member: String) throws -> Bool {
        guard let group = groupOf(member) else {
            throw TitoError.memberNotQueued(member)
        }
        return try group.markReady(member)
    }

    /// Mark every member of a queued group ready.
    @discardableResult
    func markGroupReady(_ groupID: String) throws -> Bool {
        guard let group = byID[groupID] else {
            throw TitoError.groupNotQueued(groupID)
        }
        for member in group.members {
            try group.markReady(member)
        }
        return true
    }

    /// The group a queued member belongs to, or nil if not queued.
    func groupOf(_ member: String) -> Group? {
        guard let groupID = memberIndex[member] else {
            return nil
        }
        return byID[groupID]
    }

    /// The next group that would be released, without releasing it.
    func peek() -> Group? {
        for group in order {
            if group.isReady {
                return group
            }
            if strictOrder {
                // Head of the line blocks every group behind it.
                return nil
            }
        }
        return nil
    }

    /// Release the next group that can depart, or nil.
    @discardableResult
    func release() -> Group? {
        guard let group = peek() else {
            return nil
        }
        remove(group)
        return group
    }

    /// Release every group that can currently depart, in order.
    func releaseAll() -> [Group] {
        var released: [Group] = []
        while let group = release() {
            released.append(group)
        }
        return released
    }

    /// Withdraw a waiting group, ready or not, and return it.
    @discardableResult
    func cancel(_ groupID: String) throws -> Group {
        guard let group = byID[groupID] else {
            throw TitoError.groupNotQueued(groupID)
        }
        remove(group)
        return group
    }

    /// Withdraw every waiting group, in arrival order, and return them.
    func clear() -> [Group] {
        let cleared = order
        order.removeAll()
        byID.removeAll()
        memberIndex.removeAll()
        return cleared
    }

    /// Waiting groups, in arrival order.
    var groups: [Group] {
        return order
    }

    var count: Int {
        return order.count
    }

    func contains(_ member: String) -> Bool {
        return memberIndex[member] != nil
    }

    private func remove(_ group: Group) {
        if let index = order.firstIndex(where: { $0 === group }) {
            order.remove(at: index)
        }
        byID.removeValue(forKey: group.groupID)
        for member in group.members {
            memberIndex.removeValue(forKey: member)
        }
    }
}

func flag(_ value: Bool) -> String {
    return value ? "true" : "false"
}

/// Format an optional group, as returned by peek and release.
func show(_ group: Group?) -> String {
    guard let group = group else {
        return "waiting"
    }
    return group.description
}

/// Format the group lists returned by releaseAll and clear.
func showAll(_ groups: [Group]) -> String {
    if groups.isEmpty {
        return "none"
    }
    return groups.map { $0.description }.joined(separator: ", ")
}

/// Format the ids of the waiting groups.
func showIDs(_ groups: [Group]) -> String {
    if groups.isEmpty {
        return "none"
    }
    return groups.map { $0.groupID }.joined(separator: " ")
}

/// Format a group's readiness the same way in both status lines.
func status(_ group: Group) -> String {
    return "\(group.readyCount)/\(group.count) ready [\(group.readyMembers.joined(separator: " "))]"
        + " waiting [\(group.waitingMembers.joined(separator: " "))] is_ready \(flag(group.isReady))"
}

/// Head-of-line FIFO: party-2 waits behind party-1 until it departs.
func strictOrderDemo() throws {
    print("-- strict order --")
    let queue = TitoQueue()
    print("admit \(try queue.admit("party-1", ["ann", "bob"]))")
    print("admit \(try queue.admit("party-2", ["cy"]))")
    print("len \(queue.count), groups \(showIDs(queue.groups))")
    print("contains ann \(flag(queue.contains("ann"))), contains zoe \(flag(queue.contains("zoe")))")

    guard let party1 = queue.groupOf("bob") else { return }
    print("group_of bob -> \(party1.groupID)")
    print("party-1: size \(party1.count), contains ann \(flag(party1.contains("ann"))), contains zoe \(flag(party1.contains("zoe")))")

    print("mark_ready ann -> group ready \(flag(try queue.markReady("ann")))")
    print("party-1: \(status(party1))")
    print("peek -> \(show(queue.peek()))")
    print("release -> \(show(queue.release()))")

    print("party-1.mark_ready bob -> group ready \(flag(try party1.markReady("bob")))")
    print("party-1: \(status(party1))")
    print("peek -> \(show(queue.peek()))")
    print("release -> \(show(queue.release()))")

    print("mark_group_ready party-2 -> \(flag(try queue.markGroupReady("party-2")))")
    print("release_all -> \(showAll(queue.releaseAll()))")
    print("len \(queue.count)")
}

/// Without strict order, ready party-4 leaves ahead of waiting party-3.
func relaxedOrderDemo() throws {
    print("-- relaxed order --")
    let queue = TitoQueue(strictOrder: false)
    print("strict_order \(flag(queue.strictOrder))")
    print("admit \(try queue.admit("party-3", ["dee", "eli"]))")
    print("admit \(try queue.admit("party-4", ["fay"]))")

    print("mark_group_ready party-4 -> \(flag(try queue.markGroupReady("party-4")))")
    print("peek -> \(show(queue.peek()))")
    print("release -> \(show(queue.release()))")

    print("mark_ready dee -> group ready \(flag(try queue.markReady("dee")))")
    print("release -> \(show(queue.release()))")
    print("mark_ready eli -> group ready \(flag(try queue.markReady("eli")))")
    print("release_all -> \(showAll(queue.releaseAll()))")
    print("len \(queue.count)")
}

/// Cancelling is the only way a group leaves before it is fully ready.
func cancelAndClearDemo() throws {
    print("-- cancel and clear --")
    let queue = TitoQueue()
    print("admit \(try queue.admit("party-5", ["gil", "hal"]))")
    print("admit \(try queue.admit("party-6", ["ivy"]))")
    print("admit \(try queue.admit("party-7", ["jay"]))")

    print("mark_ready gil -> group ready \(flag(try queue.markReady("gil")))")
    let cancelled = try queue.cancel("party-5")
    print("cancel party-5 -> \(cancelled) \(cancelled.readyCount)/\(cancelled.count) ready")
    print("groups \(showIDs(queue.groups))")
    print("clear -> \(showAll(queue.clear()))")
    print("len \(queue.count), contains ivy \(flag(queue.contains("ivy")))")
}

/// Run an operation that must fail, and print the reported reason.
func attempt(_ label: String, _ action: () throws -> Void) {
    do {
        try action()
        print("\(label) -> no error")
    } catch let error as TitoError {
        print("\(label) -> \(error.description)")
    } catch {
        print("\(label) -> \(error)")
    }
}

/// Every operation that would break the TITO invariants is rejected.
func errorDemo() {
    print("-- errors --")
    let queue = TitoQueue()
    do {
        print("admit \(try queue.admit("party-8", ["kim"]))")
    } catch {
        print("admit -> \(error)")
    }

    attempt("admit party-8 [kim]") { _ = try queue.admit("party-8", ["kim"]) }
    attempt("admit party-9 []") { _ = try queue.admit("party-9", []) }
    attempt("admit party-9 [jay jay]") { _ = try queue.admit("party-9", ["jay", "jay"]) }
    attempt("admit party-9 [kim]") { _ = try queue.admit("party-9", ["kim"]) }
    attempt("mark_ready zoe") { _ = try queue.markReady("zoe") }
    attempt("party-8.mark_ready zoe") {
        guard let group = queue.groupOf("kim") else { return }
        _ = try group.markReady("zoe")
    }
    attempt("mark_group_ready party-9") { _ = try queue.markGroupReady("party-9") }
    attempt("cancel party-9") { _ = try queue.cancel("party-9") }
}

do {
    try strictOrderDemo()
    print("")
    try relaxedOrderDemo()
    print("")
    try cancelAndClearDemo()
    print("")
    errorDemo()
} catch {
    print(error)
}

File: examples/tito_example.rs
// Together In Together Out - Rust example.
//
// Run: rustc -O -o /tmp/tito_rust examples/tito_example.rs && /tmp/tito_rust

use std::collections::HashMap;
use std::fmt;

/// Raised when an operation would violate the TITO invariants.
#[derive(Debug)]
struct TitoError(String);

impl fmt::Display for TitoError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl std::error::Error for TitoError {}

/// A set of members admitted together. It departs only once every member has
/// been marked ready.
struct Group {
    group_id: String,
    members: Vec<String>,
    ready: HashMap<String, bool>,
    ready_count: usize,
}

impl Group {
    fn new(group_id: &str, members: &[&str]) -> Result<Group, TitoError> {
        if members.is_empty() {
            return Err(TitoError(format!(
                "group {} must contain at least one member",
                group_id
            )));
        }
        let mut ready = HashMap::with_capacity(members.len());
        for member in members {
            if ready.insert((*member).to_string(), false).is_some() {
                return Err(TitoError(format!(
                    "group {} contains duplicate members",
                    group_id
                )));
            }
        }
        Ok(Group {
            group_id: group_id.to_string(),
            members: members.iter().map(|m| (*m).to_string()).collect(),
            ready,
            ready_count: 0,
        })
    }

    /// Members already marked ready, in admission order.
    fn ready_members(&self) -> Vec<&str> {
        self.members
            .iter()
            .filter(|m| self.ready[m.as_str()])
            .map(String::as_str)
            .collect()
    }

    /// Members not yet marked ready, in admission order.
    fn waiting_members(&self) -> Vec<&str> {
        self.members
            .iter()
            .filter(|m| !self.ready[m.as_str()])
            .map(String::as_str)
            .collect()
    }

    fn ready_count(&self) -> usize {
        self.ready_count
    }

    fn is_ready(&self) -> bool {
        self.ready_count == self.members.len()
    }

    /// Mark one member ready. Returns true once the whole group is ready.
    fn mark_ready(&mut self, member: &str) -> Result<bool, TitoError> {
        if !self.ready.contains_key(member) {
            return Err(TitoError(format!(
                "{} is not a member of group {}",
                member, self.group_id
            )));
        }
        if let Some(flag) = self.ready.get_mut(member) {
            if !*flag {
                *flag = true;
                self.ready_count += 1;
            }
        }
        Ok(self.is_ready())
    }

    fn len(&self) -> usize {
        self.members.len()
    }

    fn contains(&self, member: &str) -> bool {
        self.ready.contains_key(member)
    }
}

impl fmt::Display for Group {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{} [{}]", self.group_id, self.members.join(" "))
    }
}

/// A queue that admits and releases members group by group.
///
/// `strict_order` true (the default policy) is head-of-line FIFO: only the
/// oldest waiting group may depart. `strict_order` false lets any fully ready
/// group depart, still oldest first, so a waiting group does not block the
/// ready groups behind it.
struct TitoQueue {
    strict_order: bool,
    groups: Vec<Group>,
    member_index: HashMap<String, String>,
}

impl TitoQueue {
    fn new(strict_order: bool) -> TitoQueue {
        TitoQueue {
            strict_order,
            groups: Vec::new(),
            member_index: HashMap::new(),
        }
    }

    fn strict_order(&self) -> bool {
        self.strict_order
    }

    /// Admit members as one group and return it.
    fn admit(&mut self, group_id: &str, members: &[&str]) -> Result<&Group, TitoError> {
        let group = Group::new(group_id, members)?;
        if self.groups.iter().any(|g| g.group_id == group_id) {
            return Err(TitoError(format!(
                "group {} is already in the queue",
                group_id
            )));
        }
        let queued: Vec<&str> = group
            .members
            .iter()
            .filter(|m| self.member_index.contains_key(m.as_str()))
            .map(String::as_str)
            .collect();
        if !queued.is_empty() {
            return Err(TitoError(format!(
                "members already in the queue: {}",
                queued.join(" ")
            )));
        }
        for member in &group.members {
            self.member_index
                .insert(member.clone(), group_id.to_string());
        }
        self.groups.push(group);
        Ok(self.groups.last().expect("group was just admitted"))
    }

    /// Mark a queued member ready. Returns true once its group is ready.
    fn mark_ready(&mut self, member: &str) -> Result<bool, TitoError> {
        match self.group_of_mut(member) {
            Some(group) => group.mark_ready(member),
            None => Err(TitoError(format!("{} is not in the queue", member))),
        }
    }

    /// Mark every member of a queued group ready.
    fn mark_group_ready(&mut self, group_id: &str) -> Result<bool, TitoError> {
        let group = match self.groups.iter_mut().find(|g| g.group_id == group_id) {
            Some(group) => group,
            None => {
                return Err(TitoError(format!(
                    "group {} is not in the queue",
                    group_id
                )))
            }
        };
        let members = group.members.clone();
        for member in &members {
            group.mark_ready(member)?;
        }
        Ok(true)
    }

    /// The group a queued member belongs to, or None if not queued.
    fn group_of(&self, member: &str) -> Option<&Group> {
        let group_id = self.member_index.get(member)?;
        self.groups.iter().find(|g| &g.group_id == group_id)
    }

    fn group_of_mut(&mut self, member: &str) -> Option<&mut Group> {
        let group_id = self.member_index.get(member)?.clone();
        self.groups.iter_mut().find(|g| g.group_id == group_id)
    }

    /// The next group that would be released, without releasing it.
    fn peek(&self) -> Option<&Group> {
        self.next_index().map(|index| &self.groups[index])
    }

    /// Release the next group that can depart, or None.
    fn release(&mut self) -> Option<Group> {
        let index = self.next_index()?;
        Some(self.remove_at(index))
    }

    /// Release every group that can currently depart, in order.
    fn release_all(&mut self) -> Vec<Group> {
        let mut released = Vec::new();
        while let Some(group) = self.release() {
            released.push(group);
        }
        released
    }

    /// Withdraw a waiting group, ready or not, and return it.
    fn cancel(&mut self, group_id: &str) -> Result<Group, TitoError> {
        match self.groups.iter().position(|g| g.group_id == group_id) {
            Some(index) => Ok(self.remove_at(index)),
            None => Err(TitoError(format!(
                "group {} is not in the queue",
                group_id
            ))),
        }
    }

    /// Withdraw every waiting group, in arrival order, and return them.
    fn clear(&mut self) -> Vec<Group> {
        self.member_index.clear();
        std::mem::take(&mut self.groups)
    }

    /// Waiting groups, in arrival order.
    fn groups(&self) -> &[Group] {
        &self.groups
    }

    fn len(&self) -> usize {
        self.groups.len()
    }

    fn contains(&self, member: &str) -> bool {
        self.member_index.contains_key(member)
    }

    // The index of the next group to depart: the head if ready under strict
    // order, otherwise the oldest ready group.
    fn next_index(&self) -> Option<usize> {
        for (index, group) in self.groups.iter().enumerate() {
            if group.is_ready() {
                return Some(index);
            }
            if self.strict_order {
                // Head of the line blocks every group behind it.
                return None;
            }
        }
        None
    }

    fn remove_at(&mut self, index: usize) -> Group {
        let group = self.groups.remove(index);
        for member in &group.members {
            self.member_index.remove(member);
        }
        group
    }
}

fn flag(value: bool) -> &'static str {
    if value {
        "true"
    } else {
        "false"
    }
}

/// Format an optional group, as returned by peek and release.
fn show(group: Option<&Group>) -> String {
    match group {
        Some(group) => group.to_string(),
        None => "waiting".to_string(),
    }
}

/// Format the group lists returned by release_all and clear.
fn show_all(groups: &[Group]) -> String {
    if groups.is_empty() {
        return "none".to_string();
    }
    groups
        .iter()
        .map(Group::to_string)
        .collect::<Vec<_>>()
        .join(", ")
}

/// Format the ids of the waiting groups.
fn show_ids(groups: &[Group]) -> String {
    if groups.is_empty() {
        return "none".to_string();
    }
    groups
        .iter()
        .map(|g| g.group_id.clone())
        .collect::<Vec<_>>()
        .join(" ")
}

/// Format a group's readiness the same way in both status lines.
fn status(group: &Group) -> String {
    format!(
        "{}/{} ready [{}] waiting [{}] is_ready {}",
        group.ready_count(),
        group.len(),
        group.ready_members().join(" "),
        group.waiting_members().join(" "),
        flag(group.is_ready())
    )
}

/// Head-of-line FIFO: party-2 waits behind party-1 until it departs.
fn strict_order_demo() -> Result<(), TitoError> {
    println!("-- strict order --");
    let mut queue = TitoQueue::new(true);
    println!("admit {}", queue.admit("party-1", &["ann", "bob"])?);
    println!("admit {}", queue.admit("party-2", &["cy"])?);
    println!("len {}, groups {}", queue.len(), show_ids(queue.groups()));
    println!(
        "contains ann {}, contains zoe {}",
        flag(queue.contains("ann")),
        flag(queue.contains("zoe"))
    );

    let party1 = queue.group_of("bob").expect("party-1 is queued");
    println!("group_of bob -> {}", party1.group_id);
    println!(
        "party-1: size {}, contains ann {}, contains zoe {}",
        party1.len(),
        flag(party1.contains("ann")),
        flag(party1.contains("zoe"))
    );

    println!(
        "mark_ready ann -> group ready {}",
        flag(queue.mark_ready("ann")?)
    );
    println!(
        "party-1: {}",
        status(queue.group_of("bob").expect("party-1 is queued"))
    );
    println!("peek -> {}", show(queue.peek()));
    println!("release -> {}", show(queue.release().as_ref()));

    let ready = queue
        .group_of_mut("bob")
        .expect("party-1 is queued")
        .mark_ready("bob")?;
    println!("party-1.mark_ready bob -> group ready {}", flag(ready));
    println!(
        "party-1: {}",
        status(queue.group_of("bob").expect("party-1 is queued"))
    );
    println!("peek -> {}", show(queue.peek()));
    println!("release -> {}", show(queue.release().as_ref()));

    println!(
        "mark_group_ready party-2 -> {}",
        flag(queue.mark_group_ready("party-2")?)
    );
    println!("release_all -> {}", show_all(&queue.release_all()));
    println!("len {}", queue.len());
    Ok(())
}

/// Without strict order, ready party-4 leaves ahead of waiting party-3.
fn relaxed_order_demo() -> Result<(), TitoError> {
    println!("-- relaxed order --");
    let mut queue = TitoQueue::new(false);
    println!("strict_order {}", flag(queue.strict_order()));
    println!("admit {}", queue.admit("party-3", &["dee", "eli"])?);
    println!("admit {}", queue.admit("party-4", &["fay"])?);

    println!(
        "mark_group_ready party-4 -> {}",
        flag(queue.mark_group_ready("party-4")?)
    );
    println!("peek -> {}", show(queue.peek()));
    println!("release -> {}", show(queue.release().as_ref()));

    println!(
        "mark_ready dee -> group ready {}",
        flag(queue.mark_ready("dee")?)
    );
    println!("release -> {}", show(queue.release().as_ref()));
    println!(
        "mark_ready eli -> group ready {}",
        flag(queue.mark_ready("eli")?)
    );
    println!("release_all -> {}", show_all(&queue.release_all()));
    println!("len {}", queue.len());
    Ok(())
}

/// Cancelling is the only way a group leaves before it is fully ready.
fn cancel_and_clear_demo() -> Result<(), TitoError> {
    println!("-- cancel and clear --");
    let mut queue = TitoQueue::new(true);
    println!("admit {}", queue.admit("party-5", &["gil", "hal"])?);
    println!("admit {}", queue.admit("party-6", &["ivy"])?);
    println!("admit {}", queue.admit("party-7", &["jay"])?);

    println!(
        "mark_ready gil -> group ready {}",
        flag(queue.mark_ready("gil")?)
    );
    let cancelled = queue.cancel("party-5")?;
    println!(
        "cancel party-5 -> {} {}/{} ready",
        cancelled,
        cancelled.ready_count(),
        cancelled.len()
    );
    println!("groups {}", show_ids(queue.groups()));
    println!("clear -> {}", show_all(&queue.clear()));
    println!(
        "len {}, contains ivy {}",
        queue.len(),
        flag(queue.contains("ivy"))
    );
    Ok(())
}

/// Run an operation that must fail, and print the reported reason.
fn attempt(label: &str, result: Result<(), TitoError>) {
    match result {
        Ok(()) => println!("{} -> no error", label),
        Err(error) => println!("{} -> {}", label, error),
    }
}

/// Every operation that would break the TITO invariants is rejected.
fn error_demo() {
    println!("-- errors --");
    let mut queue = TitoQueue::new(true);
    match queue.admit("party-8", &["kim"]) {
        Ok(group) => println!("admit {}", group),
        Err(error) => println!("admit -> {}", error),
    }

    attempt(
        "admit party-8 [kim]",
        queue.admit("party-8", &["kim"]).map(|_| ()),
    );
    attempt("admit party-9 []", queue.admit("party-9", &[]).map(|_| ()));
    attempt(
        "admit party-9 [jay jay]",
        queue.admit("party-9", &["jay", "jay"]).map(|_| ()),
    );
    attempt(
        "admit party-9 [kim]",
        queue.admit("party-9", &["kim"]).map(|_| ()),
    );
    attempt("mark_ready zoe", queue.mark_ready("zoe").map(|_| ()));
    attempt(
        "party-8.mark_ready zoe",
        queue
            .group_of_mut("kim")
            .expect("kim is queued")
            .mark_ready("zoe")
            .map(|_| ()),
    );
    attempt(
        "mark_group_ready party-9",
        queue.mark_group_ready("party-9").map(|_| ()),
    );
    attempt("cancel party-9", queue.cancel("party-9").map(|_| ()));
}

fn main() -> Result<(), TitoError> {
    strict_order_demo()?;
    println!();
    relaxed_order_demo()?;
    println!();
    cancel_and_clear_demo()?;
    println!();
    error_demo();
    Ok(())
}
